data/totem.1
.\" Copyright (C) 2004 Andre Lehovich <andrel@u.arizona.edu>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License 
.\" along with this; if not write to the Free Software Foundation, Inc.
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
.TH totem 1 "2008\-08\-25" "GNOME"
.SH NAME
Totem \- GNOME desktop movie player based on GStreamer
.SH SYNOPSIS
.B totem
.RI [ option... ] " " [ filenames | URI ]
.SH DESCRIPTION
.B Totem
is a movie player for the GNOME desktop.  Through use of the
.B GStreamer
backend many popular file and streaming formats are supported including:
MPEG audio and video (MP3/MPV), Ogg, AVI, ASF, MOV, WMV,
QuickTime, and more.  Playing of DVD and VCD media is
also possible.
Although primarily intended as a movie player,
.B Totem
can also play many audio formats.
.P
For additional documentation on interactive use of 
.B Totem
select the
.I Help
menu option.  The menus allow access to some features not
available through keybindings or command-line options.
.SH OPTIONS
.TP
.B \-?, --help
Output a longer help message then quit.  (Any discrepancies
between this man page and the output of
.B totem --help
should be reported as a documentation bug.)
.TP
.B --version
Output version information then quit.
.P
The following options command an already-running instance of
.B Totem
to do something; they are useful for remote-control scripting. If 
.B Totem
is not already running, these commands will launch a new instance of
.B Totem
but will not do anything further.
.TP
.B --play-pause
Tell an already-running instance of 
.B Totem
to toggle between play and pause.
.TP
.B --play
Tell an already-running instance of
.B Totem
to play (has no effect if already playing)
.TP
.B --pause
Tell an already-running instance of
.B Totem
to pause (has no effect if already paused)
.TP
.B --next
Tell an already-running instance of 
.B Totem
to skip to the next movie or chapter in the playlist.
.TP
.B --previous
Tell an already-running instance of 
.B Totem
to return to the previous movie or chapter in the playlist.
.TP
.B --seek-fwd
Tell an already-running instance of 
.B Totem
to seek forwards 15 seconds in the current movie.
.TP
.B --seek-bwd
Tell an already-running instance of 
.B Totem
to seek backwards 15 seconds in the current movie.
.TP
.B --volume-up
Tell an already-running instance of 
.B Totem
to raise the volume by 8%.
.TP
.B --volume-down
Tell an already-running instance of 
.B Totem
to lower the volume by 8%.
.TP
.B --mute
Tell an already-running instance of 
.B Totem
to toggle between muted and unmuted audio.
.TP
.B --fullscreen
Tell an already-running instance of 
.B Totem
to toggle fullscreen mode.
.TP
.B --toggle-controls
Tell an already-running instance of
.B Totem
to toggle showing the controls.
.TP
.B --quit
Tell an already-running instance of
.B Totem
to quit.
.TP
.BI "--enqueue " filename|URI
Tell an already-running instance of Totem to add a new stream
to the playlist.
.TP
.BI "--replace " filename|URI
Tell an already-running instance of Totem to play 
from the playlist.
.SH KEYBINDINGS
Global keybindings for
.BR Totem :
.P
.ta \w'Down-arrow   'u	
a	cycle between aspect ratios
.br
p	toggle between play and pause
.br
Esc	exit full screen mode
.br
f	toggle full screen
.br
h	toggle display of on-screen controls
.br
0	resize window to 50% original size
.br
1	resize window to 100% original size
.br
2	resize window to 200% original size
.br
.br
r	zoom in the video
.br
t	zoom out the video
.br
d	start and stop the telestrator (drawing) mode
.br
e	erase the drawing
.br
Left-arrow        skip back 15 seconds
.br
Right-arrow       skip forward 60 seconds
.br
Shift+Left-arrow  skip back 5 seconds
.br
Shift+Right-arrow skip forward 15 seconds
.br
Ctrl+Left-arrow   skip back 3 minutes
.br
Ctrl+Right arrow  skip forward 10 minutes
.br
Up-arrow	increase volume by 8%
.br
Down-arrow	decrease volume by 8%
.br
b		jump back to previous chapter/movie in playlist
.br
n		jump to next chapter/movie in playlist
.br
q		quit
.br
Ctrl+E          eject the playing optical media
.br
Ctrl+O		open a new file
.br
Ctrl+L		open a new URI
.br
F9		toggle display of the playlist
.br
m		show the DVD menu
.br
c		show the DVD chapter menu
.SH FILES
.I /usr/lib/totem/plugins/
.RS
The global plugins directory, which contains all the system-installed
plugins.
.RE
.I ~/.local/share/totem/plugins/
.RS
The user plugins directory, containing plugins available only to that user.
It is always in the "totem/plugins" subdirectory of your 
.B XDG_DATA_HOME
directory.
.RE
.I ~/.config/totem/state.ini
.RS
The configuration file for the main window's appearance, stored in a
simple INI format.
.RE
.SH BUGS
.B Totem
cannot play files using some proprietary codecs.
.SH AUTHOR
.B Totem
was written by Bastien Nocera <hadess@hadess.net>.
This manual page was written by Andre Lehovich for the
Debian Project.
.SH DISTRIBUTION
The latest version of Totem may be downloaded from
.UR http://projects.gnome.org/totem/
<http://projects.gnome.org/totem/>
.UE
.SH SEE ALSO
.BR "totem-video-thumbnailer" (1),
.P
The online documentation available through the program's
.I Help
menu.


NEWS
New features and significant updates in version...

Major changes in 3.10.1:
- Fix possible abort on startup
- Fix warning on startup

Major changes in 3.10.0:
- Updated translations

Major changes in 3.9.92:
- RTL bug fixes for playback icons
- Fix year tag not showing in some cases
- Remove obsolete xplayer-disc usage
- Port zeitgeist plugin to zeitgeist-2.0

Major changes in 3.9.5:
- Add work-around to allow streaming of Vimeo videos
- Don't hide the titlebar on fullscreen (as we still have
  a menu bar)
- Make thumbnailer work even when VA-API support is enabled
- Avoid warnings from the rotation plugin for unsupported locations
- Fix aspect ratio in dual-head setups
- Fix playback when video location is not mounted
- Access streams directly rather than through FUSE, making
  buffering and playback smoother
- Fix excessive redraws when using rotation, and make
  rotation and crop work together

Major changes in 3.8.2:
- Work-around problem with initial window size
- Fix errors in gallery name print when filename contains "&"
- Fix translations not being applied in nautilus' properties
- Fix Python Console not working
- Fix screenshot keyboard shortcut

Major changes in 3.8.0:
- Updated translations

Major changes in 3.7.93:
- Fix incorrect Vala generated files getting in the release

Major changes in 3.7.92 from 3.6.3:

- Use yelp-tools instead of gnome-doc-utils
- Add YouTube format configuration for grilo
- Fix possible Unicode errors with MPRIS plugin
- Use libgd's icon view for search results
- Fix warnings for some files in the browse sidebar
- Fix possible crashes in the chapters plugin
- Fix playing audio CD tracks launched from Nautilus
- Fix running when clutter-gst is in a different prefix

Plugins:
- Remove use of deprecated widgets in subtitles downloader

Major changes in 3.5.90:

- Plenty of build fixes to make it possible to build with Wayland support
- Increase thumbnail size to 256 px

Movie player:
- Rename to "Videos"
- Hide titlebar when maximised
- Move some menu items to the app menu
- Command-line parsing bug fixes
- Use Grilo to list DVDs and other optical media
- Remove connection speed preferences
- Stop DVD playback when disc is ejected
- Fix ordering of the playlist after drag'n'drop

Backend:
- Port to GStreamer 1.0
- Add HTTP proxy support
- Much improved error messages
- Add variable rate playback
- Use Clutter for the OSD
- Disable visualisations by default
- Animate rotations
- Add av-offset support
- Prepare the ground for chapter support in non-DVDs

Plugins:
- Make screenshot UI-less, like in gnome-shell
- Simplify the screensaver and browser plugin code
- Remove publish and iPlayer plugins
- Use Grilo 0.2
- Better search sidebar
- Add hook for site-specific user-agents
- Add hook for custom subtitle selection through plugins
- Move recent-files handling to a plugin

3.4.0:
* Mark some strings as translatable
* Handle Clutter failing to initialise
* Don't handle Ctrl+Space (which is used for IM switching)

3.3.92:
* Check plugins for introspection in configure

3.3.90:
* Show Vorbis DESCRIPTION field in Comments
* Use GTK+ code to detect the workspace's work area
* Do use direct seeking for files in the wastebasket
* Thumbnail FLAC files too
* Fix crash when using MPRIS plugin
* Fix compilation of the rotation plugin with newer Vala

3.3.4:
* Re-enable and fix compilation of zeitgeist-dp plugin
* Use new GLib threading API if compiling with GLib ≥ 2.31.0
* Fix some problems with `make check` on the Python plugins

3.3.3:
* Add plugin to replace the Flash plugin for online videos
* Add support for RTMP URLs
* Fix vala plugins compilation
* Fix --time option in thumbnailer
* Fix possible crashes with the browser plugin
  on OpenBSD
* Remove obsolete youtube and bemused plugins
* Port browser plugin to GDBus
* Fix getting the cover art in the movie player
* Fix crasher when Xplayer is passed arguments

3.2.1:
* Fix crash when framerate isn't available
* Quieten the thumbnailer when thumbnailing non-videos
* Loads of translations

3.2.0:
* Fix Python console settings conversion bug
* Loads of updated translations

3.1.92:
* Mention our support of a few more mime-types
* Handle HTTP Live Streaming (in the main player,
  and the browser plugin)
* Loads of translations

3.1.91:
* Port D-Bus service to MPRIS 2.1
* Store and restore the rotation status of videos
* Fix the "save" button not work in the playlist
* Screensaver fixes for non-GNOME sessions
* Port to pygobject 3.0

3.1.4:
* Add Grilo plugin
* Remove Coherence UPNP, Tracker and Jamendo plugins
* Add plugin to rotate videos

* Fix audio tracks not appearing with their language names
* Add keyboard shortcut for "Save"
* Remove sidebar drop-down item, and use menu items instead
* Fix translations of the nautilus properties window
* Subtitles plugin bug fixes

3.1.0:
* Use clutter-gst to draw the video, the
  only major related change is simpler zoom modes
* Use GStreamer API directly for audio preview
  utility (used in nautilus), thumbnailer and
  nautilus properties window
* Show container format in the properties
* Add ability to switch video tracks
* Add ability to "thumbnail" DVD ISOs
* Advertise support for audio/x-stm and audio/x-s3m
* Fix possible dropped chapters in chapter plugin

* Code:
  - Port media-player-keys plugin to GDBus
  - Add pylint support for all Python plugins, and fix
    errors from it, and port all of the plugins to
    GObject Introspection
  - Simplify screenshot code

* Removed:
  - Remove XVidMode support in backend
  - Remove video indexer, it wasn't used
  - Remove thumbnail plugin which doesn't serve a purpose
    under GNOME 3

3.0.0:
* Fix packing and spacing in the properties
* Fix subtitle downloader and jamendo plugins
* Loads more translations

2.91.93:
* Fix activation of all Python plugins
* Make Python console work again
* Loads of translations

2.91.92:
* Plenty of screensaver related bug fixes
* Make "Open Location" dialogue modal

src/totem.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/* 
 * Copyright (C) 2001-2007 Bastien Nocera <hadess@hadess.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
 *
 * The Totem project hereby grant permission for non-gpl compatible GStreamer
 * plugins to be used and distributed together with GStreamer and Totem. This
 * permission are above and beyond the permissions granted by the GPL license
 * Totem is covered by.
 *
 * Monday 7th February 2005: Christian Schaller: Add exception clause.
 * See license_change file for details.
 *
 */

#include "config.h"

#include <glib.h>
#include <glib-object.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <locale.h>
#include <string.h>

#ifdef GDK_WINDOWING_X11
/* X11 headers */
#include <X11/Xlib.h>
#endif

#include "totem.h"
#include "totem-private.h"
#include "totem-interface.h"
#include "totem-options.h"
#include "totem-menu.h"
#include "totem-session.h"
#include "totem-uri.h"
#include "totem-preferences.h"
#include "totem-rtl-helpers.h"
#include "totem-sidebar.h"
#include "video-utils.h"

static gboolean startup_called = FALSE;

/* Debug log message handler: discards debug messages unless Totem is run with TOTEM_DEBUG=1.
 * If we're building in the source tree, enable debug messages by default. */
static void
debug_handler (const char *log_domain,
               GLogLevelFlags log_level,
               const char *message,
               GSettings *settings)
{
	static int debug = -1;

	if (debug < 0)
		debug = g_settings_get_boolean (settings, "debug");

	if (debug)
		g_log_default_handler (log_domain, log_level, message, NULL);
}

static void
set_rtl_icon_name (Totem *totem,
		   const char *widget_name,
		   const char *icon_name)
{
	GtkAction *action;

	action = GTK_ACTION (gtk_builder_get_object (totem->xml, widget_name));
	gtk_action_set_icon_name (action, totem_get_rtl_icon_name (icon_name));
}

static void
app_init (Totem *totem, char **argv)
{
	GtkSettings *gtk_settings;
	char *sidebar_pageid;

	if (gtk_clutter_init (NULL, NULL) != CLUTTER_INIT_SUCCESS)
		g_warning ("gtk-clutter failed to initialise, expect problems from here on.");

	gtk_settings = gtk_settings_get_default ();
	g_object_set (G_OBJECT (gtk_settings), "gtk-application-prefer-dark-theme", TRUE, NULL);

	/* Debug log handling */
	g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG, (GLogFunc) debug_handler, totem->settings);

	/* Main window */
	totem->xml = totem_interface_load ("totem.ui", TRUE, NULL, totem);
	if (totem->xml == NULL)
		totem_action_exit (NULL);

	set_rtl_icon_name (totem, "play", "media-playback-start");
	set_rtl_icon_name (totem, "next-chapter", "media-skip-forward");
	set_rtl_icon_name (totem, "previous-chapter", "media-skip-backward");
	set_rtl_icon_name (totem, "skip-forward", "media-seek-forward");
	set_rtl_icon_name (totem, "skip-backwards", "media-seek-backward");

	totem->win = GTK_WIDGET (gtk_builder_get_object (totem->xml, "totem_main_window"));

	/* Menubar */
	totem_ui_manager_setup (totem);

	/* The sidebar */
	playlist_widget_setup (totem);

	/* The rest of the widgets */
	totem->state = STATE_STOPPED;
	totem->seek = GTK_WIDGET (gtk_builder_get_object (totem->xml, "tmw_seek_hscale"));
	totem->seekadj = gtk_range_get_adjustment (GTK_RANGE (totem->seek));
	totem->volume = GTK_WIDGET (gtk_builder_get_object (totem->xml, "tmw_volume_button"));
	totem->statusbar = GTK_WIDGET (gtk_builder_get_object (totem->xml, "tmw_statusbar"));
	totem->seek_lock = FALSE;
	totem->fs = totem_fullscreen_new (GTK_WINDOW (totem->win));
	gtk_scale_button_set_adjustment (GTK_SCALE_BUTTON (totem->fs->volume),
					 gtk_scale_button_get_adjustment (GTK_SCALE_BUTTON (totem->volume)));
	gtk_range_set_adjustment (GTK_RANGE (totem->fs->seek), totem->seekadj);

	totem_session_setup (totem, argv);
	totem_setup_file_monitoring (totem);
	totem_setup_file_filters ();
	totem_callback_connect (totem);

	sidebar_pageid = totem_setup_window (totem);

	/* Show ! */
	if (optionstate.fullscreen == FALSE) {
		gtk_widget_show (totem->win);
		totem_gdk_window_set_waiting_cursor (gtk_widget_get_window (totem->win));
	} else {
		gtk_widget_realize (totem->win);
	}

	totem->controls_visibility = TOTEM_CONTROLS_UNDEFINED;

	/* Show ! (again) the video widget this time. */
	video_widget_create (totem);
	gtk_widget_grab_focus (GTK_WIDGET (totem->bvw));
	totem_fullscreen_set_video_widget (totem->fs, totem->bvw);

	if (optionstate.fullscreen != FALSE) {
		gtk_widget_show (totem->win);
		gdk_flush ();
		totem_action_fullscreen (totem, TRUE);
	}

	/* The prefs after the video widget is connected */
	totem->prefs_xml = totem_interface_load ("preferences.ui", TRUE, NULL, totem);

	totem_setup_preferences (totem);

	totem_setup_recent (totem);

	/* Command-line handling */
	totem_options_process_late (totem, &optionstate);

	/* Initialise all the plugins, and set the default page, in case
	 * it comes from a plugin */
	totem_object_plugins_init (totem);
	totem_sidebar_set_current_page (totem, sidebar_pageid, FALSE);
	g_free (sidebar_pageid);

	if (totem->session_restored != FALSE) {
		totem_session_restore (totem, optionstate.filenames);
	} else if (optionstate.filenames != NULL && totem_action_open_files (totem, optionstate.filenames)) {
		totem_action_play_pause (totem);
	} else {
		totem_action_set_mrl (totem, NULL, NULL);
	}

	/* Set the logo at the last minute so we won't try to show it before a video */
	bacon_video_widget_set_logo (totem->bvw, "totem");

	if (optionstate.fullscreen == FALSE)
		gdk_window_set_cursor (gtk_widget_get_window (totem->win), NULL);

	gtk_window_set_application (GTK_WINDOW (totem->win), GTK_APPLICATION (totem));
}

static void
app_startup (GApplication *application,
		Totem        *totem)
{
	/* We don't do anything here, as we need to know the options
	 * when we set everything up.
	 * Note that this will break D-Bus activation of the application */
	startup_called = TRUE;
}

static int
app_command_line (GApplication             *app,
		  GApplicationCommandLine  *command_line,
		  Totem                    *totem)
{
	GOptionContext *context;
	int argc;
	char **argv;

	argv = g_application_command_line_get_arguments (command_line, &argc);

	/* Reset the options, if they were used before */
	memset (&optionstate, 0, sizeof (optionstate));

	/* Options parsing */
	context = totem_options_get_context ();
	g_option_context_set_help_enabled (context, FALSE);
	if (g_option_context_parse (context, &argc, &argv, NULL) == FALSE) {
	        g_option_context_free (context);
	        return 1;
	}
	g_option_context_free (context);

	totem_options_process_early (totem, &optionstate);

	/* Don't create another window if we're remote.
	 * We can't use g_application_get_is_remote() because it's not registered yet */
	if (startup_called != FALSE) {
		app_init (totem, argv);

		gdk_notify_startup_complete ();

		/* Don't add files again through totem_options_process_for_server() */
		g_strfreev (optionstate.filenames);
		optionstate.filenames = NULL;
		startup_called = FALSE;
	}

	/* Now do something with it */
	totem_options_process_for_server (totem, &optionstate);

	g_strfreev (argv);
	return 0;
}

int
main (int argc, char **argv)
{
	Totem *totem;

	setlocale (LC_ALL, "");
	bindtextdomain (GETTEXT_PACKAGE, GNOMELOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

#ifdef GDK_WINDOWING_X11
	if (XInitThreads () == 0)
	{
		gtk_init (&argc, &argv);
		g_set_application_name (_("Videos"));
		totem_action_error_and_exit (_("Could not initialize the thread-safe libraries."), _("Verify your system installation. Totem will now exit."), NULL);
	}
#endif

	g_type_init ();

	g_set_prgname ("totem");
	g_set_application_name (_("Videos"));
	gtk_window_set_default_icon_name ("totem");
	g_setenv("PULSE_PROP_media.role", "video", TRUE);


	/* Build the main Totem object */
	totem = g_object_new (TOTEM_TYPE_OBJECT,
			      "application-id", "org.gnome.Totem",
			      "flags", G_APPLICATION_HANDLES_COMMAND_LINE,
			      NULL);
	totem->settings = g_settings_new (TOTEM_GSETTINGS_SCHEMA);

	g_signal_connect (G_OBJECT (totem), "startup",
			  G_CALLBACK (app_startup), totem);
	g_signal_connect (G_OBJECT (totem), "command-line",
			  G_CALLBACK (app_command_line), totem);

	g_application_run (G_APPLICATION (totem), argc, argv);

	return 0;
}
